### PR TITLE
fix: dockerfile and ci improvements

### PR DESCRIPTION
**Pull request by jonaro00**
## Description of change

- Cleanup and upgrades in Dockerfiles and CI.
- Minor fixes in builder code.

## How has this been tested? (if applicable)

Tested some deploys with a minimal service.

### CODE REVIEW

**Review by iulianbarbu — COMMENTED** [review]
Looks good @jonaro00 ! Only have a few comments with regards to some features that might become throw away work pretty soon (1/2 weeks timeframe at most). We can either merge the PR without them or wait a bit merging it until we'll be able to test it (and by that time we'll probably be really close with the the new builder). I think the best thing we can do is to address the comments and avoid merging the additional `sscache` and `protoc` from the deployer. What do you think?

File: deployer/prepare.sh [hunk 1]
```diff
@@ -62,12 +62,18 @@ replace-with = "shuttle-crates-io-mirror"' >> $CARGO_HOME/config.toml
...
 # The image should already have these: https://github.com/docker-library/buildpack-deps/blob/65d69325ad741cea6dee20781c1faaab2e003d87/debian/buster/Dockerfile
 apt update
 apt install -y llvm-dev libclang-dev clang cmake
+
+# Install protoc since some users may need it
+ARCH="linux-x86_64" && \
+VERSION="22.2" && \
+curl -OL "https://github.com/protocolbuffers/protobuf/releases/download/v$VERSION/protoc-$VERSION-$ARCH.zip" && \
+    unzip -o "protoc-$VERSION-$ARCH.zip" bin/protoc "include/*" -d /usr/local && \
+    rm -f "protoc-$VERSION-$ARCH.zip"
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/prepare.sh:79]: These lines were just moved further down in the file. But yes, builder service will make it irrelevant later.

File: deployer/prepare.sh [hunk 2]
```diff
@@ -30,26 +32,24 @@ if [[ $PROD != "true" ]]; then
...
-cargo install shuttle-runtime --path "/usr/src/shuttle/runtime" --bin shuttle-next --features next
+# Set up sccache for build caching
+export SCCACHE_VERSION='v0.5.3'
+curl -L https://github.com/mozilla/sccache/releases/download/$SCCACHE_VERSION/sccache-$SCCACHE_VERSION-x86_64-unknown-linux-musl.tar.gz \
+  | tar -xOz sccache-$SCCACHE_VERSION-x86_64-unknown-linux-musl/sccache \
+  > /usr/local/cargo/bin/sccache \
+  && chmod +x /usr/local/cargo/bin/sccache
+echo '
+[build]
+rustc-wrapper = "/usr/local/cargo/bin/sccache"' >> $CARGO_HOME/config.toml
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/prepare.sh:48]: Yeah, I knew that this would probably be throw-away. I'll exclude it from this PR and look for other improvements.

**Comment by jonaro00** [comment]
I found a major optimization in caching in the Containerfile when building deployer.
I moved the handling of prepare.sh to before any shuttle code is copied, which means it is not re-run every time you change the code.
For me, this took the docker build time of "change 1 line -> make shuttle-deployer" from 4m20s to 40s! 😄

**Review by chesedo — APPROVED** [review]
LGTM, thanks @jonaro00!!

File: Containerfile [hunk 3]
```diff
@@ -1,47 +1,64 @@
...
-FROM docker.io/library/rust:${RUSTUP_TOOLCHAIN}-buster as shuttle-build
-RUN apt-get update &&\
-    apt-get install -y curl
+#syntax=docker/dockerfile-upstream:1.4
+
 
-RUN cargo install cargo-chef
+# Base image for builds and cache
+ARG RUSTUP_TOOLCHAIN
+FROM rust:${RUSTUP_TOOLCHAIN}-buster as shuttle-build
```
Comment (oddgrd) [inline Containerfile:6]: Perhaps we should keep this? For reference: https://github.com/shuttle-hq/shuttle/pull/723
```suggestion
FROM docker.io/library/rust:${RUSTUP_TOOLCHAIN}-buster as shuttle-build
```

File: Containerfile [hunk 4]
```diff
@@ -1,47 +1,64 @@
...
+RUN apt update && apt install -y curl
```
Comment (oddgrd) [inline Containerfile:7]: Do we still need curl here btw? :thinking:

**Review by oddgrd — APPROVED** [review]
Looks good to me two! :partying_face: Just two very minor questions

File: Containerfile (hunk 4 above)
Comment (jonaro00) [reply Containerfile]: Moved to deployer. Only used there to install protoc.

Comment (oddgrd) [reply Containerfile]: Cool, thanks!

**Pull request merged**